ReactiveCocoa/Swift/SignalProducer.swift

```diff
@@ -107,42 +107,57 @@ public struct SignalProducer<T, E: ErrorType> {
 	public static func buffer(_ capacity: Int = Int.max) -> (SignalProducer, Signal<T, E>.Observer) {
 		precondition(capacity >= 0)
 
-		let lock = NSRecursiveLock()
+		let lock = NSLock()
 		lock.name = "org.reactivecocoa.ReactiveCocoa.SignalProducer.buffer"
 
 		var events: [Event<T, E>] = []
 		var terminationEvent: Event<T,E>?
-		var observers: Bag<Signal<T, E>.Observer>? = Bag()
+		let observers: Atomic<Bag<Signal<T, E>.Observer>?> = Atomic(Bag())
 
 		let producer = self { observer, disposable in
 			lock.lock()
+
+			var token: RemovalToken?
+			observers.modify { (var observers) in
+				token = observers?.insert(observer)
+				return observers
+			}
+
 			for event in events {
 				observer.put(event)
 			}
+
 			if let terminationEvent = terminationEvent {
 				observer.put(terminationEvent)
 			}
 
-			let token = observers?.insert(observer)
 			lock.unlock()
 
 			if let token = token {
 				disposable.addDisposable {
-					lock.lock()
-					observers?.removeValueForToken(token)
-					lock.unlock()
+					observers.modify { (var observers) in
+						observers?.removeValueForToken(token)
+						return observers
+					}
 				}
 			}
 		}
 
 		let observer = Signal<T, E>.Observer { event in
```

Comment:
Can you rename this sink. It'll help distinguish it even further from the observer in the above closure.